src/services/blocks/BlocksService.ts

```diff
@@ -450,84 +473,122 @@ export class BlocksService extends AbstractService {
-						degree: c.degree.toNumber(),
-						negative: c.negative,
-					};
-				}
-			);
-
-			const parentParentHash: Hash = await this.getParentParentHash(
-				api,
-				parentHash,
-				block
-			);
-
-			const [version, multiplier] = await Promise.all([
-				api.rpc.state.getRuntimeVersion(parentParentHash),
-				api.query.transactionPayment.nextFeeMultiplier.at(parentHash),
-			]);
-
-			[specName, specVersion] = [
-				version.specName.toString(),
-				version.specVersion.toNumber(),
-			];
-
-			runtimeDoesNotMatch =
-				specName !== api.runtimeVersion.specName.toString() ||
-				specVersion !== api.runtimeVersion.specVersion.toNumber();
-
-			if (runtimeDoesNotMatch) {
-				const metadata = await api.rpc.state.getMetadata(parentParentHash);
-
-				decorated = expandMetadata(api.registry, metadata);
-			}
-
-			calcFee = CalcFee.from_params(
-				coefficients,
-				multiplier.toString(10),
-				perByte.toString(10),
+			// This particular runtime version is not supported with fee calcs or
+			// does not have the necessay materials to build calcFee
+			return {
+				specVersion,
 				specName,
-				specVersion
-			);
+			};
 		}
 
+		// Now that we know the exact runtime supports fee calcs, make sure we have
+		// the weights in the store
+		this.blockWeightStore[specVersion] ||= await this.getWeight(
+			api,
+			parentHash
+		);
+
+		const calcFee = CalcFee.from_params(
+			coefficients,
+			multiplier.toString(10),
+			perByte.toString(10),
+			specName,
+			specVersion
+		);
+
 		return {
 			calcFee,
 			specName,
 			specVersion,
-			decorated,
-			runtimeDoesNotMatch,
 		};
 	}
 
+	/**
+	 *	Get a formatted blockweight store value for the runtime corresponding to the given block hash.
+	 *
+	 * @param api ApiPromise
+	 * @param blockHash Hash of a block in the runtime to get the extrinsic base weight(s) for
+	 * @returns formatted block weight store entry
+	 */
+	private async getWeight(
+		api: ApiPromise,
+		blockHash: BlockHash
+	): Promise<WeightValue> {
+		const metadata = await api.rpc.state.getMetadata(blockHash);
+		const {
+			consts: { system },
+		} = expandMetadata(api.registry, metadata);
+
+		let weightValue;
+		if (((system.blockWeights as unknown) as BlockWeights)?.perClass) {
+			const {
+				normal,
+				operational,
+				mandatory,
+			} = ((system.blockWeights as unknown) as BlockWeights)?.perClass;
+
+			const perClass = {
+				normal: {
+					baseExtrinsic: normal.baseExtrinsic.toBigInt(),
+				},
+				operational: {
+					baseExtrinsic: operational.baseExtrinsic.toBigInt(),
+				},
+				mandatory: {
+					baseExtrinsic: mandatory.baseExtrinsic.toBigInt(),
+				},
+			};
+
+			weightValue = { perClass };
+		} else if (system.extrinsicBaseWeight) {
+			weightValue = {
+				extrinsicBaseWeight: ((system.extrinsicBaseWeight as unknown) as AbstractInt).toBigInt(),
+			};
+		} else {
+			throw new InternalServerError(
+				'Could not find a extrinsic base weight in metadata'
+			);
```

Comment:
@TarikGul  @emostov is throwing errors the kosher way in `sidecar` to return errors? Not a fan of `try-catch` as you know and it'd be better if we had a proper error type we use consistently, but I'm might see this with Rust-colored glasses. 
If this is how we do things typically I'll shut up about it (for now). :)

Reply:
I think moving forward having our own internal Error type and, Error handler would be a great improvement. I do think the idea holds a lot of merit. Currently this is how things are done, but that being said there is room for improvement. I have seen monkey-patched `try-catch` implementation's for error handling that have improved performance in some blogs, so this some good food for thought.

Reply:
Yes, the errors are explicitly caught inside of the services with

https://github.com/paritytech/substrate-api-sidecar/blob/415f0302b05aeac013c93227ac19e4928427206a/src/controllers/AbstractController.ts#L77-L88

Once express 5.0 stabilizes this wrapper for requests will no longer be needed since it will catch async throws out of the box.

Also important to keep in mind that certain types of error formats are explicitly supported by sidecar's middleware: https://github.com/paritytech/substrate-api-sidecar/tree/master/src/middleware/error - this includes errors from the HttpError library that you see here.

I was not aware of any penalties for `try catch` when our wrote our error handling pipeline. With a quick google search, the first stake overflow answer seems to indicated that `try catch` is no slower than any other code since node v8.3 (we require min v12): https://github.com/paritytech/substrate-api-sidecar/tree/master/src/middleware/error

However probably warrants a little further investigation just to make sure there are not some low hanging optimizations.

Reply:
> Once express 5.0 stabilizes this wrapper for requests will no longer be needed since it will catch async throws out of the box.

Didn't know about that improvement is coming in 5.0. Pretty awesome!

Reply:
Cool, I rest my case.

(That is, my main beef with `try-catch` and exceptions in general is that one can never be really sure where errors are caught without reading through the whole reverse call-chain. They call this "spooky action at a distance" I think.)